### ISSUE TITLE
NaN gradients appear with a combination of normalization and masking

### ISSUE THREAD

**Issue by hrbigelow** [google/jax#17070]
### Description

The following happens on Colab CPU and T4 GPU, and GTX 1070 MaxQ (laptop).

Test here:  [Colab Example](https://colab.research.google.com/drive/1lfzAeHajMrzeCOPKs6g8kz5QoO6_FMPv?usp=sharing)

In the below code, the function `main_fn` accepts logits and a `where` mask.  It performs a softmax on the logits using the mask.  It then computes entropy of the resulting distribution, also using the where mask.  

It can run in two modes, `curse=True` or `curse=False`.  If `curse==True`, it additionally performs (an unnecessary) normalization of the softmax output.

If the `where` mask has some False entries, and `curse == True`, then some gradients are `NaN`.  In these cases, the function itself gives the expected output.

This is just a toy example, but in other examples, I am first summing multiple softmax distributions together, and then normalizing the result.  This is why I need the normalize step.

Any ideas greatly appreciated!

```python
import jax
import jax.numpy as jnp

def main_fn(logits, where, curse):
    """
    logit: t
    where: t
    returns: ()
    """
    attn = jax.nn.softmax(logits, axis=0, where=where, initial=-1e6) # t
    if curse:
        attn = attn * jnp.sum(attn) ** -1
    return jnp.sum(attn * -jnp.log2(attn), where=where, initial=1.0)

if __name__ == '__main__':

    grad_fn = jax.grad(main_fn)

    logits = jnp.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0])
    where1 = logits < 4.0
    where2 = logits < 10.0  # everywhere

    for where in (where1, where2):
        for curse in (True, False):
            val = main_fn(logits, where, curse)
            grad = grad_fn(logits, where, curse)
            print(f'where: {where.astype(jnp.int32).tolist()}, '
                    f'curse: {curse:4d}, '
                f'val: {val.astype(jnp.float16).tolist():10f}, '
                f'grad: {grad.astype(jnp.float16).tolist()}')

"""
where: [1, 1, 1, 0, 0, 0], curse:    1, val:   2.201172, grad: [nan, nan, nan, 0.0, 0.0, 0.0]
where: [1, 1, 1, 0, 0, 0], curse:    0, val:   2.201172, grad: [0.20458984375, 0.203125, -0.40771484375, 0.0, 0.0, 0.0]
where: [1, 1, 1, 1, 1, 1], curse:    1, val:   2.476562, grad: [0.027313232421875, 0.0574951171875, 0.1107177734375, 0.17724609375, 0.1456298828125, -0.5185546875]
where: [1, 1, 1, 1, 1, 1], curse:    0, val:   2.476562, grad: [0.027313232421875, 0.0574951171875, 0.1107177734375, 0.17724609375, 0.1456298828125, -0.5185546875]
"""
```

### What jax/jaxlib version are you using?

jax v0.4.12, jaxlib v0.4.12+cuda11.cudnn86

### Which accelerator(s) are you using?

GTX 1070 MaxQ, CPU, Colab T4 GPU

### Additional system info

Linux 96ae20233979 5.15.0-78-generic

### NVIDIA GPU info

```
Thu Aug 10 21:42:31 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.125.06   Driver Version: 525.125.06   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:01:00.0 Off |                  N/A |
| N/A   43C    P8     3W /  30W |      6MiB /  8192MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

**Comment by hrbigelow** [comment]
Closing this issue.  I'll reopen a simpler example.
